### PR TITLE
update vlad, gmm, kmeans & sift

### PR DESCRIPTION
**Pull request by hubutui**
1. use fused types for np.float32 and np.float64 support
2. update test cases for vlad
3. update documents

### CODE REVIEW

File: cyvlfeat/vlad/tests/vlad_test.py [hunk 1]
```diff
@@ -1,53 +1,89 @@
...
+        expected_enc = simple_vlad(x.transpose(), mu.transpose(), assignments.transpose())
+        expected_enc = np.sign(expected_enc) * np.sqrt(np.abs(expected_enc))
+        expected_enc = expected_enc / np.linalg.norm(expected_enc)
+        observed_enc = vlad(x, mu, assignments, square_root=True)
+        assert_allclose(observed_enc, expected_enc)
+
+def test_individual():
+    for dtype in [np.float32, np.float64]:
+        x, mu, assignments = get_fake_data(dtype=dtype)
+        expected_enc = simple_vlad(x.transpose(), mu.transpose(), assignments.transpose())
```
Comment (patricksnape) [inline cyvlfeat/vlad/tests/vlad_test.py:60]: Minor thing just to say that you can use `.T` as a shorthand for `.transpose()`

File: cyvlfeat/vlad/tests/vlad_test.py [hunk 2]
```diff
@@ -1,53 +1,89 @@
...
+        assignments = np.eye(3, dtype=dtype)
+        observed_enc = vlad(x, mu, assignments, unnormalized=True)
+        expected_enc = np.array([1, 2, 3]).astype(dtype)
+        assert_allclose(expected_enc, observed_enc)
+
+def test_rand():
+    for dtype in [np.float32, np.float64]:
+        x, mu, assignments = get_fake_data(dtype)
+        observed_enc = vlad(x, mu, assignments, unnormalized=True)
+        expected_enc = simple_vlad(x.transpose(), mu.transpose(), assignments.transpose())
```
Comment (patricksnape) [inline cyvlfeat/vlad/tests/vlad_test.py:37]: Also just to note that transposing the matrices every time will cause a copy of the data matrices due to the call to `ascontiguous` in the methods (which may cause surprising behavior for users when they get out of memory errors).

**Comment by patricksnape** [comment]
Awesome - great work

**Comment by patricksnape** [comment]
Minor comment is just that this is the reason the other methods were flipped to avoid the transpose -> because the transpose will cause a copy

File: cyvlfeat/vlad/tests/vlad_test.py (hunk 2 above)
Comment (hubutui) [reply cyvlfeat/vlad/tests/vlad_test.py:37]: Oh, this code will only be run once when we test, just wanna make sure the python binding work as expected.

Comment (patricksnape) [reply cyvlfeat/vlad/tests/vlad_test.py:37]: Sure - I just meant that if people are building their matrices like that then they may be surprised

**Pull request merged**